From peterz@infradead.org  Wed Mar 11 13:30:19 2009
From: Peter Zijlstra <peterz@infradead.org>
Date: Mon, 02 Mar 2009 13:21:17 +0100
Subject: Staging: irq: remove IRQF_DISABLED
Message-ID: <1235996477.5330.174.camel@laptop>


People are playing odd games with IRQF_DISABLED, remove it.

Its not reliable, since shared interrupt lines could disable it for you,
and its possible and allowed for archs to disable IRQs to limit IRQ nesting.

Therefore, simply mandate that _ALL_ IRQ handlers are run with IRQs disabled.

[ This _should_ not break anything, since we've mandated that IRQ handlers
  _must_ be able to deal with this for a _long_ time ]

IRQ handlers should be fast, no if buts and any other exceptions. We also have
plenty instrumentation to find any offending IRQ latency sources.

Signed-off-by: Peter Zijlstra <a.p.zijlstra@chello.nl>
Acked-by: Ingo Molnar <mingo@elte.hu>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/staging/me4000/me4000.c           |    6 +++---
 drivers/staging/meilhaus/me0600_ext_irq.c |    6 +-----
 drivers/staging/meilhaus/me1400_ext_irq.c |    6 +-----
 drivers/staging/meilhaus/me4600_ai.c      |    6 +-----
 drivers/staging/meilhaus/me4600_ao.c      |    6 +-----
 drivers/staging/meilhaus/me4600_ext_irq.c |    6 +-----
 drivers/staging/meilhaus/me6000_ao.c      |    6 +-----
 drivers/staging/meilhaus/me8100_di.c      |    6 +-----
 drivers/staging/meilhaus/me8200_di.c      |   12 ++----------
 drivers/staging/meilhaus/me8200_do.c      |    6 +-----
 10 files changed, 13 insertions(+), 53 deletions(-)

--- a/drivers/staging/me4000/me4000.c
+++ b/drivers/staging/me4000/me4000.c
@@ -842,7 +842,7 @@ static int alloc_ao_contexts(struct me40
 			/* Request the interrupt line */
 			err =
 			    request_irq(ao_context->irq, me4000_ao_isr,
-					IRQF_DISABLED | IRQF_SHARED,
+					IRQF_SHARED,
 					ME4000_NAME, ao_context);
 			if (err) {
 				printk(KERN_ERR
@@ -1555,7 +1555,7 @@ static int me4000_open(struct inode *ino
 		/* Request the interrupt line */
 		err =
 		    request_irq(ext_int_context->irq, me4000_ext_int_isr,
-				IRQF_DISABLED | IRQF_SHARED, ME4000_NAME,
+				IRQF_SHARED, ME4000_NAME,
 				ext_int_context);
 		if (err) {
 			printk(KERN_ERR
@@ -3032,7 +3032,7 @@ static int me4000_ai_prepare(struct me40
 		/* Request the interrupt line */
 		err =
 		    request_irq(ai_context->irq, me4000_ai_isr,
-				IRQF_DISABLED | IRQF_SHARED, ME4000_NAME,
+				IRQF_SHARED, ME4000_NAME,
 				ai_context);
 		if (err) {
 			printk(KERN_ERR
--- a/drivers/staging/meilhaus/me0600_ext_irq.c
+++ b/drivers/staging/meilhaus/me0600_ext_irq.c
@@ -438,11 +438,7 @@ me0600_ext_irq_subdevice_t *me0600_ext_i
 	subdevice->irq = irq;
 
 	err = request_irq(subdevice->irq, me0600_isr,
-#ifdef IRQF_DISABLED
-			  IRQF_DISABLED | IRQF_SHARED,
-#else
-			  SA_INTERRUPT | SA_SHIRQ,
-#endif
+			  IRQF_SHARED,
 			  ME0600_NAME, (void *)subdevice);
 
 	if (err) {
--- a/drivers/staging/meilhaus/me1400_ext_irq.c
+++ b/drivers/staging/meilhaus/me1400_ext_irq.c
@@ -463,11 +463,7 @@ me1400_ext_irq_subdevice_t *me1400_ext_i
 	subdevice->irq = irq;
 
 	err = request_irq(irq, me1400_ext_irq_isr,
-#ifdef IRQF_DISABLED
-			  IRQF_DISABLED | IRQF_SHARED,
-#else
-			  SA_INTERRUPT | SA_SHIRQ,
-#endif
+			  IRQF_SHARED,
 			  ME1400_NAME, (void *)subdevice);
 
 	if (err) {
--- a/drivers/staging/meilhaus/me4600_ai.c
+++ b/drivers/staging/meilhaus/me4600_ai.c
@@ -307,11 +307,7 @@ me4600_ai_subdevice_t *me4600_ai_constru
 	// Register interrupt service routine.
 	subdevice->irq = irq;
 	if (request_irq(subdevice->irq, me4600_ai_isr,
-#ifdef IRQF_DISABLED
-			IRQF_DISABLED | IRQF_SHARED,
-#else
-			SA_INTERRUPT | SA_SHIRQ,
-#endif
+			IRQF_SHARED,
 			ME4600_NAME, subdevice)) {
 		PERROR("Cannot register interrupt service routine.\n");
 		me_subdevice_deinit((me_subdevice_t *) subdevice);
--- a/drivers/staging/meilhaus/me4600_ao.c
+++ b/drivers/staging/meilhaus/me4600_ao.c
@@ -2581,11 +2581,7 @@ me4600_ao_subdevice_t *me4600_ao_constru
 	if (subdevice->fifo) {
 		subdevice->irq = irq;
 		if (request_irq(subdevice->irq, me4600_ao_isr,
-#ifdef IRQF_DISABLED
-				IRQF_DISABLED | IRQF_SHARED,
-#else
-				SA_INTERRUPT | SA_SHIRQ,
-#endif
+				IRQF_SHARED,
 				ME4600_NAME, subdevice)) {
 			PERROR("Cannot get interrupt line.\n");
 			PDEBUG("free circ_buf = %p size=%d",
--- a/drivers/staging/meilhaus/me4600_ext_irq.c
+++ b/drivers/staging/meilhaus/me4600_ext_irq.c
@@ -425,11 +425,7 @@ me4600_ext_irq_subdevice_t *me4600_ext_i
 	subdevice->irq = irq;
 
 	if (request_irq(subdevice->irq, me4600_ext_irq_isr,
-#ifdef IRQF_DISABLED
-			IRQF_DISABLED | IRQF_SHARED,
-#else
-			SA_INTERRUPT | SA_SHIRQ,
-#endif
+			IRQF_SHARED,
 			ME4600_NAME, subdevice)) {
 		PERROR("Cannot register interrupt.\n");
 		kfree(subdevice);
--- a/drivers/staging/meilhaus/me6000_ao.c
+++ b/drivers/staging/meilhaus/me6000_ao.c
@@ -2639,11 +2639,7 @@ me6000_ao_subdevice_t *me6000_ao_constru
 	if (subdevice->fifo & ME6000_AO_HAS_FIFO) {
 		subdevice->irq = irq;
 		if (request_irq(subdevice->irq, me6000_ao_isr,
-#ifdef IRQF_DISABLED
-				IRQF_DISABLED | IRQF_SHARED,
-#else
-				SA_INTERRUPT | SA_SHIRQ,
-#endif
+				IRQF_SHARED,
 				ME6000_NAME, subdevice)) {
 			PERROR("Cannot get interrupt line.\n");
 			PDEBUG("free circ_buf = %p size=%d",
--- a/drivers/staging/meilhaus/me8100_di.c
+++ b/drivers/staging/meilhaus/me8100_di.c
@@ -637,11 +637,7 @@ me8100_di_subdevice_t *me8100_di_constru
 	/* Register interrupt service routine. */
 	subdevice->irq = irq;
 	err = request_irq(subdevice->irq, me8100_isr,
-#ifdef IRQF_DISABLED
-			  IRQF_DISABLED | IRQF_SHARED,
-#else
-			  SA_INTERRUPT | SA_SHIRQ,
-#endif
+			  IRQF_SHARED,
 			  ME8100_NAME, (void *)subdevice);
 
 	if (err) {
--- a/drivers/staging/meilhaus/me8200_di.c
+++ b/drivers/staging/meilhaus/me8200_di.c
@@ -817,19 +817,11 @@ me8200_di_subdevice_t *me8200_di_constru
 	subdevice->irq = irq;
 	if (subdevice->version > 0) {	// NEW
 		err = request_irq(subdevice->irq, me8200_isr_EX,
-#ifdef IRQF_DISABLED
-				  IRQF_DISABLED | IRQF_SHARED,
-#else
-				  SA_INTERRUPT | SA_SHIRQ,
-#endif
+				  IRQF_SHARED,
 				  ME8200_NAME, (void *)subdevice);
 	} else {		//OLD
 		err = request_irq(subdevice->irq, me8200_isr,
-#ifdef IRQF_DISABLED
-				  IRQF_DISABLED | IRQF_SHARED,
-#else
-				  SA_INTERRUPT | SA_SHIRQ,
-#endif
+				  IRQF_SHARED,
 				  ME8200_NAME, (void *)subdevice);
 	}
 
--- a/drivers/staging/meilhaus/me8200_do.c
+++ b/drivers/staging/meilhaus/me8200_do.c
@@ -560,11 +560,7 @@ me8200_do_subdevice_t *me8200_do_constru
 
 	/* Request the interrupt line */
 	err = request_irq(irq, me8200_do_isr,
-#ifdef IRQF_DISABLED
-			  IRQF_DISABLED | IRQF_SHARED,
-#else
-			  SA_INTERRUPT | SA_SHIRQ,
-#endif
+			  IRQF_SHARED,
 			  ME8200_NAME, (void *)subdevice);
 
 	if (err) {
